docs/cs-proof.md

```diff
@@ -0,0 +1,800 @@
+
+### Multiplication gates
+
+Each multiplication gate takes two input variables and multiplies them to get an output.
+That relation for all multiplication gates is represented by \\(n\\) constraints:
+
+\\[
+\mathbf{a}\_L \circ \mathbf{a}\_R = \mathbf{a}\_O,
+\\]
+where:
+* \\(\mathbf{a}\_L\\) is the vector of the first input to each gate,
+* \\(\mathbf{a}\_R\\) is the vector of the second input to each gate,
+* \\(\mathbf{a}\_O\\) is the vector of multiplication results.
+
+### Linear constraints
+
+Linear constraints are expressed using a vector of \\(q\\) equations that use linear combinations of the [variables](#variables):
+
+\\[
+\mathbf{W}\_L \cdot \mathbf{a}\_L +
+\mathbf{W}\_R \cdot \mathbf{a}\_R +
+\mathbf{W}\_O \cdot \mathbf{a}\_O =
+\mathbf{W}\_V \cdot \mathbf{v} +
+\mathbf{c},
+\\]
+where:
+
+* \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O\\) are weights applied to the respective inputs and outputs of the [multiplication gates](#multiplication-gates) (low-level [variables](#variables)),
+* \\(\mathbf{W}\_V\\) are weights applied to the high-level [variables](#variables) \\(\mathbf{v}\\),
+* \\(\mathbf{c}\\) is a vector of constant terms used in the linear constraints.
+
+
+### Building constraints
+
+Bulletproofs framework allows building constraint systems _on the fly_, without a trusted setup.
+This allows instantiating constraints from a _family of constraint systems_ parametrized by
+public values and commitments to the secret inputs.
+As a result, the instantiation can be thought of as a _challenge_ generated by a verifier to a prover.
+
+The prover starts out by committing to its secret inputs \\(\mathbf{v}\\)
+and obtaining \\(m\\) _variables_ representing these inputs.
+
+Then, the prover creates linear combinations of available [variables](#variables)
+
+The prover performs a combination of the following operations to generate the constraints:
+
+1. **Allocate a multiplier:** a new [multiplication gate](#multiplication-gates) is added represented by three [low-level variables](#variables) \\(a_L, a_R, a_O\\), for left input, right input and output value respectively.
+2. **Add a constraint:** a [linear combination](#linear-constraints) of any number of [variables](#variables) is encoded into appropriate positions in matrices \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O, \mathbf{W}\_V\\) and a vector of constants \\(\mathbf{c}\\).
+3. **Request a challenge scalar:** a random scalar returned in response to committed [high-level variables](#variables).
+
+
+### Unconstrained variables
+
+Often a constraint system is composed of _gadgets_: subsets of constraints
+that implement features of a higher-level protocol.
+The “wires” that connect the gadgets are called _uncommitted variables_.
+They are created from left and right variables \\(a\_L, a\_R\\) of
+additional multiplication gates.
+Output variables \\(a\_O\\) are intentionally left unconstrained.
```

Comment:
This is a confusing sentence - a_O variables are not actually left "unconstrained" since they are actually constrained (implicitly) to be the multiplication of a_L and a_R. Maybe a clearer way to explain this is that the output variables a_O are intentionally not given as options for unconstrained variables, since they are already implicitly constrained against a_L and a_R.

Reply:
good point. How about:

> Output variables `a_O` are not exposed as they are already constrained by a multiplication gate.

Reply:
Or rather:

> Output variables `a_O` cannot be used as unconstrained variables since they are implicitly constrained by a multiplication gate, and are not provided to the user by the constraint system.

Reply:
Actually, rewrote like this:

>### Uncommitted variables
>
>Often a constraint system is composed of _gadgets_: subsets of constraints
>that implement features of a higher-level protocol.
>The “wires” that connect the gadgets are called _uncommitted variables_.
>
>Uncommitted variables are created from left and right variables \\(a\_L, a\_R\\) of additional multiplication gates.
>Output variables \\(a\_O\\) are implicitly constrained by a [multiplication gate](#multiplication-gates),
>so they cannot be used as independent uncommitted variables.